### PR TITLE
feat(das) add shrex.Sub integration into DASer

### PR DESCRIPTION
**Pull request by walldiss**
## Overview

Besides integrations, it allows recent headers to be sampled right away, without need to wait next available worker to process the job. This way recent headers are not restricted by concurrency limit and it should reduce sampling latency to allow Broadcast happen faster.

Closes https://github.com/celestiaorg/celestia-node/issues/1466
Closes https://github.com/celestiaorg/celestia-node/issues/1718

TODO:
 - [x] create issue for removal of rate limit restrictions for recent headers
 - [ ] need shrex.Sub fx integration (probably another PR)
 
  

## Checklist

- [x] New and updated code has appropriate documentation
- [x] New and updated code has new and/or updated testing
- [x] Required CI checks are passing
- [x] Visual proof for any user facing features like CLI or documentation updates
- [x] Linked issues closed with keywords

### CODE REVIEW

File: das/options.go [hunk 1]
```diff
@@ -97,6 +98,14 @@ func (p *Parameters) Validate() error {
 		)
 	}
 
+	// SampleTimeout = 0 would fail every sample operation with timeout error
+	if p.MaxRecentHeaderGap < 1 {
+		return errInvalidOptionValue(
+			"SampleTimeout",
```
Comment (distractedm1nd) [inline das/options.go:104]: typo

File: das/options.go [hunk 2]
```diff
@@ -97,6 +98,14 @@ func (p *Parameters) Validate() error {
 		)
 	}
 
+	// SampleTimeout = 0 would fail every sample operation with timeout error
```
Comment (distractedm1nd) [inline das/options.go:101]: typo

File: das/coordinator.go [hunk 3]
```diff
@@ -4,6 +4,8 @@ import (
 	"context"
 	"sync"
 
+	"github.com/celestiaorg/celestia-node/share/p2p/shrexsub"
+
```
Comment (distractedm1nd) [inline das/coordinator.go:8]:
```suggestion
```

File: das/coordinator_test.go [hunk 4]
```diff
@@ -8,6 +8,9 @@ import (
 	"testing"
 	"time"
 
+	"github.com/celestiaorg/celestia-node/share"
+	"github.com/celestiaorg/celestia-node/share/p2p/shrexsub"
+
```
Comment (distractedm1nd) [inline das/coordinator_test.go:13]:
```suggestion
```

File: das/worker.go [hunk 5]
```diff
@@ -71,12 +76,23 @@ func (w *worker) run(
...
-		} else {
-			log.Debugw("sampled header", "height", h.Height(), "hash", h.Hash(),
-				"square width", len(h.DAH.RowsRoots), "data root", h.DAH.Hash(), "finished (s)", time.Since(startSample))
+			continue
+		}
+
+		log.Debugw("sampled header", "height", h.Height(), "hash", h.Hash(),
+			"square width", len(h.DAH.RowsRoots), "data root", h.DAH.Hash(), "finished (s)", time.Since(startSample))
+
+		// notify network about recent header being available to be synced via shrex.Getter
```
Comment (distractedm1nd) [inline das/worker.go:108]:
```suggestion
		// notify network via shrexsub that we are available as a peer for syncing the recent data
		// light nodes provide a stub `shrexsub.BroadcastFn`, as they do not run shrexsub 
```
@renaynay and others please help find a good comment, don't think mine is great either

**Review by distractedm1nd — CHANGES_REQUESTED** [review]

File: das/worker.go [hunk 6]
```diff
@@ -85,13 +85,25 @@ func (w *worker) run(
...
+				"data root", h.DAH.Hash(),
+				"err", err,
+			)
 			continue
 		}
 
-		log.Debugw("sampled header", "height", h.Height(), "hash", h.Hash(),
-			"square width", len(h.DAH.RowsRoots), "data root", h.DAH.Hash(), "finished (s)", time.Since(startSample))
+		log.Debugw(
+				"sampled header",
```
Comment (distractedm1nd) [inline das/worker.go:100]: there is a lint error here you may have missed from the merge

File: das/state.go [hunk 7]
```diff
@@ -122,19 +123,15 @@ func (s *coordinatorState) nextJob() (next job, found bool) {
 	return j, true
 }
 
-func (s *coordinatorState) nextFromPriority() (job, bool) {
-	for len(s.priority) > 0 {
-		next := s.priority[len(s.priority)-1]
-		s.priority = s.priority[:len(s.priority)-1]
+func (s *coordinatorState) nextFromRetry() (job, bool) {
+	for len(s.retry) == 0 {
```
Comment (renaynay) [inline das/state.go:127]:
```suggestion
	if len(s.retry) == 0 {
```

File: das/worker.go (hunk 5 above)
Comment (renaynay) [reply das/worker.go:108]:
```suggestion
		// notify network about availability of new block data (note: only full nodes can notify)
```

**Review by distractedm1nd — APPROVED** [review]

**Review by renaynay — APPROVED** [review]

**Review by vgonkivs — APPROVED** [review]

**Pull request merged**